Add generic phrase details page

sophiebrown01 opened the pull request:
This PR adds the generic phrase detail page. Dates are hardcoded at present until we pass them in from the search page and links need to be removed or pointed at pages which currently don't exist, however the main features of the page have been implemented:

- Number of survey responses with this generic phrase (chart)
- Most frequent exact matches
- Most frequent co-occurring generic phrases
- How users mentions this phrase

To-do:

- Remove hardcoded dates for date filters
- Take top n "Most frequent co-occurring generic phrases"
- Remove links or create the pages to point them to

Commit:
Add skeleton generic phrase detail page

Commit:
Add most frequent for generic phrase to phrase model

Commit:
Add survey responses by generic phrase to Mention model and hook up

Commit:
Fix spec helper to include additional helpers

Commit:
Add survey answers by generic phrase to SurveyAnswer model and hook-up to front-end

Commit:
Add co-occurring generic phrases

This commit adds co-occurring generic phrases to the generic phrase detail page. This lists all other generic phrases that the selected generic phrase appears alongside in survey answers, ordered by most frequent.

Commit:
Remove spec helper method

Commit:
Show only top 10 most frequent co-occurring generic phrases

Commit:
Hook up date filter from search page to detail page

Commit:
Add most frequent generic phrases to trending

Commit:
Only show top 10 exact phrases on generic phrase detail page

oscarwyatt commented on app/helpers/generic_phrases_helper.rb:
```diff
@@ -28,57 +28,48 @@ def generic_phrase_table_headers(presenter)
-        value: adj,
-        selected: adj == presenter.adjective,
-      }
+  def map_most_frequent_exact_matches_to_table(presenter)
```
As I said in some other review, I think things like this belong in the presenter but we will sort that out at some other time

sophiebrown01 replied:
To be fair, more of it is in the presenter now than it was before. I think too that `create_options_for_select_list` should belong in a helper as it's pretty generic.

oscarwyatt commented on app/helpers/generic_phrases_helper.rb:
```diff
@@ -28,57 +28,48 @@ def generic_phrase_table_headers(presenter)
-      link = link_to generic_phrase, generic_phrase_path(id: generic_phrase_id)
-
+  def map_most_frequent_co_occurring_generic_phrases_to_table(presenter)
+    presenter.most_frequent_co_occurring_generic_phrases.map do |k, v|
```
Not super keen on variables like `k` `v` but as above, we may move this code anyway so we can change the names to be something meaningful then

sophiebrown01 replied:
Will change this

oscarwyatt commented on app/models/concerns/generic_phrase_concern.rb:
```diff
@@ -0,0 +1,16 @@
+    def co_occurring_generic_phrases(generic_phrases_for_survey_answers)
+      generic_phrase_combinations = []
+
+      generic_phrases_for_survey_answers.group_by(&:first).each do |_, r|
```
I'd suggest using `each_with_object([])` as it's slightly tidier. I don't know what an `r` is, is there a more meaningful variable name that it could have? 💭

oscarwyatt replied:
However, nice one for putting this straight into a  module 💯

sophiebrown01 replied:
Will rename this to be more descriptive, thanks for pointing out

oscarwyatt commented on app/models/generic_phrase.rb:
```diff
@@ -22,4 +23,42 @@ def self.search(start_date, end_date, options = {})
+
+    generic_phrase_combinations = co_occurring_generic_phrases(generic_phrases_for_survey_answers)
+    generic_phrase_combinations
+      .each_with_object({}) { |(a, b), hsh| hsh["#{a}-#{b}"] = generic_phrase_combinations.count([a, b]) }
```
Again, variables `a` and `b` require me to look backwards in the code to understand what they are, increasing mental load, if there is something that is more meaningful that'd be good

sophiebrown01 replied:
Spot on

oscarwyatt commented on app/views/generic_phrases/show.html.erb:
```diff
@@ -0,0 +1,173 @@
+  <div class="govuk-grid-column-two-thirds">
+    <div class="table-group">
+      <h2 class="govuk-heading-m">Most frequent co-occurring generic phrases</h2>
+      <%#= debug @presenter.most_frequent_co_occurring_generic_phrases %>
```
Guessing this was included by accident? Or does it serve some purpose?

oscarwyatt commented on app/views/generic_phrases/show.html.erb:
```diff
@@ -0,0 +1,173 @@
+          text: "Content trigger warning: some feedback can be distressing."
+      } %>
+
+      <% @presenter.survey_answers.each do |sa| %>
```
Sorry, I know this is all dreadfully dull but although `sa` is ok..for anyone new to the codebase `survey_answer` is more descriptive and clearer

sophiebrown01 replied:
I must have been in my SQL-aliasing frame of mind when writing this

oscarwyatt commented on app/controllers/generic_phrases_controller.rb:
```diff
@@ -35,6 +43,28 @@ def adjective_results
 
+  def most_frequent_exact_matches(generic_phrase)
+    Phrase
+      .most_frequent_for_generic_phrase(generic_phrase, Date.new(2020, 4, 1), Date.new(2020, 4, 10))
```
The `searchable` module makes it easy to get these dates from the date picker, it's set to the last 7 days by default  if it's not included in the page parameters https://github.com/alphagov/govuk-user-intent-survey-explorer/pull/42/files#diff-cc6e703301c6ed33065f186cec0d2939R24

sophiebrown01 replied:
I hardcoded dates here for the purposes of development and not needing to constantly change the dates to data that matches. I'll change them to use your searchable module now, which I was aware of previously.

oscarwyatt reviewed (changes requested):
Great work. Mainly really minor things about variable names

Commit:
Make code more legible

This commit makes code more legible, updating variable names where it can otherwise be hard to understand what they represent and removing links on the Generic Phrase show page where there are currently no pages to link to.

oscarwyatt reviewed (approved):
Lovin 'yo work

Commit:
Add contents section
